server/market/balancer.go

```diff
@@ -77,43 +139,123 @@ func (b *DEXBalancer) CheckBalance(acctAddr string, assetID uint32, qty, lots ui
 	}
+	if bal == 0 {
+		log.Tracef("(*DEXBalancer).CheckBalance(%q, %d, %d, %d, %d) false for zero balance",
+			acctAddr, assetID, qty, lots, redeems)
+		return false
+	}
+
+	// Set up fee tracking for tokens.
+	var feeID uint32 // Asset ID of the base chain asset.
+	// feeQty will track the fee asset order-locked quantity, but is only used
+	// for tokens. When not a token (when assetID = base chain), the
+	// order-locked quantity is summed in reqFunds as for any other asset. We'll
+	// fetch feeBal to catch any zero bals or other errors early before
+	// iterating markets and querying the Swapper.
+	var feeQty, feeBal uint64 // Current balance.
+	feeBalancer := backedAsset.feeBalancer
+	isToken := feeBalancer != nil
+	if isToken {
+		feeID = feeBalancer.assetInfo.ID
+		feeBal, err = feeBalancer.balancer.AccountBalance(acctAddr)
+		if err != nil {
+			log.Error("(*DEXBalancer).CheckBalance: error getting fee asset balance for %q: %v", acctAddr, err)
+			return false
+		}
+		if feeBal == 0 {
+			log.Tracef("(*DEXBalancer).CheckBalance(%q, %d, %d, %d, %d) false for zero fee asset (%s) balance",
+				acctAddr, assetID, qty, lots, redeems, feeBalancer.assetInfo.Symbol)
+			return false
+		}
+	}
+
+	var swapFees, redeemFees uint64
+	addFees := func(assetInfo *dex.Asset, l uint64, r int) {
+		// The fee rate assigned to redemptions is at the discretion of the
+		// user. MaxFeeRate is used as a conservatively high estimate. This is
+		// then a server policy that clients must satisfy.
+		redeemFees += uint64(r) * assetInfo.RedeemSize * assetInfo.MaxFeeRate
+		swapFees += calc.RequiredOrderFunds(0, 0, l, assetInfo) // Alt: assetInfo.SwapSize * l
+	}
+
+	// Add the fees for the requested lots.
+	addFees(backedAsset.assetInfo, lots, redeems)
+
+	// Prepare a function to add fees pending orders for the asset and each
+	// asset in the fee-family.
+	addPending := func(assetID uint32) (q uint64) {
+		ba, found := b.assets[assetID]
+		if !found {
+			log.Errorf("(*DEXBalancer).CheckBalance: asset ID %d not a configured backedBalancer", assetID)
+			return 0
+		}
 
-	// Add quantity for unfilled orders.
-	for _, mt := range b.tunnels {
-		newQty, newLots, newRedeems := mt.AccountPending(acctAddr, assetID)
-		lots += newLots
-		qty += newQty
-		redeems += newRedeems
+		var l uint64
+		var r int
+		for _, mt := range ba.markets {
+			newQty, newLots, newRedeems := mt.AccountPending(acctAddr, assetID)
+			l += newLots
+			q += newQty
+			r += newRedeems
+		}
+
+		// Add in-process swaps.
+		newQty, newLots, newRedeems := b.matchNegotiator.AccountStats(acctAddr, assetID)
+		l += newLots
+		q += newQty
+		r += newRedeems
+
+		addFees(ba.assetInfo, l, r)
+		return
+	}
+
+	qty += addPending(assetID)
+
+	// Add fee-family asset pending fees.
+	for famID := range backedAsset.feeFamily {
+		if q := addPending(famID); isToken && famID == feeID {
+			// Add the quantity of the base chain order reserves to our fee
+			// asset balance check.
+			feeQty = q
```

Comment:
Depending on the intent here (I'm not completely studied yet), but would this need to be either a `+=` or left as `=` but with a break?

Reply:
> Edit: oh, are you after the side effects of addPending regardless of famID?

Yes.